File: backend/app/modules/ai/api/ai.py
import uuid
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.api.dependencies import (
    get_ai_service,
    get_config_service,
    get_current_user,
    get_user_context,
)
from app.core.database import get_db
from app.core.entities.system import User
from app.core.exceptions import NotFoundError, ValidationError
from app.core.services.config import ConfigService
from app.modules.ai.services.ai import AIService

router = APIRouter()

class AskAureonRequest(BaseModel):
    context_type: str  # "signal" or "recommendation"
    context_id: uuid.UUID
    question: str

class AIFeedbackRequest(BaseModel):
    generation_id: uuid.UUID
    rating: int  # 1 (thumbs up) or -1 (thumbs down)
    comment: Optional[str] = None

@router.post("/ai/global")
def generate_global_briefing(
    current_user: User = Depends(get_current_user),
    ai_service: AIService = Depends(get_ai_service)
):
    try:
        return ai_service.generate_briefing("global", user_id=current_user.id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/ai/weekly")
def generate_weekly_briefing(
    current_user: User = Depends(get_current_user),
    ai_service: AIService = Depends(get_ai_service)
):
    try:
        return ai_service.generate_briefing("weekly", user_id=current_user.id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/ai/monthly")
def generate_monthly_briefing(
    current_user: User = Depends(get_current_user),
    ai_service: AIService = Depends(get_ai_service)
):
    try:
        return ai_service.generate_briefing("monthly", user_id=current_user.id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/ai/qa")
def ask_aureon(
    req: AskAureonRequest,
    current_user: User = Depends(get_current_user),
    ai_service: AIService = Depends(get_ai_service)
):
    try:
        response, generation_id = ai_service.ask_aureon(req.context_type, req.context_id, req.question, user_id=current_user.id)
        return {"response": response, "generation_id": str(generation_id)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/ai/feedback")
def submit_ai_feedback(
    req: AIFeedbackRequest,
    current_user: User = Depends(get_current_user),
    ai_service: AIService = Depends(get_ai_service)
):
    try:
        feedback = ai_service.submit_feedback(req.generation_id, req.rating, req.comment, user_id=current_user.id)
        return {"id": str(feedback.id), "generation_id": str(feedback.generation_id), "rating": feedback.rating}
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e.message)) from e
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e.message)) from e

@router.post("/ai/recommendations/{recommendation_id}/explain")
def explain_recommendation(
    recommendation_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    ai_service: AIService = Depends(get_ai_service)
):
    try:
        return ai_service.explain_recommendation(recommendation_id, user_id=current_user.id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# --- Bare analytics endpoints, single-user facade via get_user_context ---

@router.get("/analytics/ai/briefings")
def fetch_briefing_history(
    limit: int = Query(30),
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
    ai_service: AIService = Depends(get_ai_service),
):
    get_user_context(db, user)
    return ai_service.get_briefing_history(limit=limit)

@router.get("/analytics/ai/single/{symbol}")
@router.post("/analytics/ai/single/{symbol}")
def get_ai_take(
    symbol: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
    ai_service: AIService = Depends(get_ai_service)
):
    get_user_context(db, user)  # ensures default Portfolio exists
    return ai_service.get_single_asset_take(symbol, user_id=user.id)

@router.get("/analytics/ai/usage")
def get_ai_usage(
    since: Optional[datetime] = None,
    until: Optional[datetime] = None,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
    ai_service: AIService = Depends(get_ai_service),
):
    get_user_context(db, user)
    return ai_service.get_usage_summary(since, until)

@router.post("/analytics/ai/news/batch")
def analyze_news_batch(
    config_svc: ConfigService = Depends(get_config_service),
):
    task_id = config_svc.dispatch_job("fetch_news")
    return {"status": "queued", "message": "News batch queued", "task_id": task_id}

File: backend/src/routes/ai/ai.ts
import { Router, type Request, type Response, type NextFunction } from "express";
import { prisma } from "../../prisma";
import { getCurrentUser, getUserContext } from "../../lib/users";
import { requireUuidParam } from "../../lib/validation";
import { RequestValidationError } from "../../lib/errors";
import {
  generateBriefing,
  askAureon,
  explainRecommendation,
  submitFeedback,
  getBriefingHistory,
  getSingleAssetTake,
  getUsageSummary,
} from "../../lib/ai/aiService";
import { dispatchJob } from "../../lib/settings/jobDispatch";

// Port of app/modules/ai/api/ai.py — the briefing/Q&A/feedback/explain
// surface, briefing-history/single-asset-take/usage-summary analytics
// endpoints, and news-batch dispatch.
export const aiRouter = Router();

function parseOptionalDateQuery(value: unknown, paramName: string): Date | null {
  if (value === undefined) return null;
  const d = new Date(String(value));
  if (Number.isNaN(d.getTime())) {
    throw new RequestValidationError(`${paramName} must be a valid ISO 8601 datetime`);
  }
  return d;
}

aiRouter.post("/ai/global", async (req, res, next) => {
  try {
    const user = await getCurrentUser();
    res.json(await generateBriefing("global", user.id));
  } catch (e) {
    next(e);
  }
});

aiRouter.post("/ai/weekly", async (req, res, next) => {
  try {
    const user = await getCurrentUser();
    res.json(await generateBriefing("weekly", user.id));
  } catch (e) {
    next(e);
  }
});

aiRouter.post("/ai/monthly", async (req, res, next) => {
  try {
    const user = await getCurrentUser();
    res.json(await generateBriefing("monthly", user.id));
  } catch (e) {
    next(e);
  }
});

aiRouter.post("/ai/qa", async (req, res, next) => {
  try {
    const { context_type: contextType, context_id: contextId, question } = req.body ?? {};
    if (typeof contextType !== "string" || typeof contextId !== "string" || typeof question !== "string") {
      throw new RequestValidationError("context_type, context_id, and question are required");
    }
    requireUuidParam(contextId, "context_id");
    const user = await getCurrentUser();
    const { response, generationId } = await askAureon(contextType, contextId, question, user.id);
    res.json({ response, generation_id: generationId });
  } catch (e) {
    next(e);
  }
});

aiRouter.post("/ai/feedback", async (req, res, next) => {
  try {
    const { generation_id: generationId, rating, comment } = req.body ?? {};
    requireUuidParam(generationId, "generation_id");
    const user = await getCurrentUser();
    const feedback = await submitFeedback(generationId, Number(rating), comment ?? null, user.id);
    res.json(feedback);
  } catch (e) {
    next(e);
  }
});

aiRouter.post("/ai/recommendations/:id/explain", async (req, res, next) => {
  try {
    requireUuidParam(req.params.id, "recommendation_id");
    const user = await getCurrentUser();
    res.json(await explainRecommendation(req.params.id, user.id));
  } catch (e) {
    next(e);
  }
});

aiRouter.get("/analytics/ai/briefings", async (req, res, next) => {
  try {
    const limit = req.query.limit ? Number(req.query.limit) : 30;
    res.json(await getBriefingHistory(limit));
  } catch (e) {
    next(e);
  }
});

// Port of get_ai_take. get_user_context is a side-effect-only call in Python
// (ensures the default Portfolio exists; the returned id is unused here).
async function handleSingleAssetTake(req: Request, res: Response, next: NextFunction) {
  try {
    const user = await getCurrentUser();
    await prisma.$transaction((tx) => getUserContext(tx));
    res.json(await getSingleAssetTake(String(req.params.symbol), user.id));
  } catch (e) {
    next(e);
  }
}
aiRouter.get("/analytics/ai/single/:symbol", handleSingleAssetTake);
aiRouter.post("/analytics/ai/single/:symbol", handleSingleAssetTake);

aiRouter.get("/analytics/ai/usage", async (req, res, next) => {
  try {
    await getCurrentUser();
    await prisma.$transaction((tx) => getUserContext(tx));
    const since = parseOptionalDateQuery(req.query.since, "since");
    const until = parseOptionalDateQuery(req.query.until, "until");
    res.json(await getUsageSummary(since, until));
  } catch (e) {
    next(e);
  }
});

// Port of analyze_news_batch — pure job dispatch, no AI logic of its own.
aiRouter.post("/analytics/ai/news/batch", async (req, res, next) => {
  try {
    const taskId = await dispatchJob("fetch_news");
    res.json({ status: "queued", message: "News batch queued", task_id: taskId });
  } catch (e) {
    next(e);
  }
});
